requestSite should return a rejected promise on error

jsnajdr opened the pull request:
This PRs improves the `requestSite` action to return more reasonable values when being dispatched:
- when the site request succeeds, the action returns a promise that resolves to the `site` object.
- when the site request fails, the action returns a rejected promise with the error.

Additionally, when the REST request returns a Jetpack JSON API module error, the request is re-tried with a `force: 'wpcom'` parameter that serves the information from the shadow site and avoid sending a request to the actual Jetpack site.

The action also won't store the `site` into Redux state if the user doesn't have manage capabilities there. But the `site` will be returned in the promise, as there are many use cases that use it.

The previous version of `requestSite` never returned a rejected promise, and resolved either to `undefined` in case of success, and to a `SITE_REQUEST_FAILURE` action object in case of error or unmanaged site. That's rather confusing IMO.

Several usages of `dispatch( requestSite( id ) ).then( ... )` need to be modified to handle the rejected promise with an empty `catch` handler. In these cases, we wait until the `requestSite` action completes, either successfully or not, and then figure out the result by looking at Redux state for the requested site.

This PR was inspired by reviewing #52979 by @Aurorum and seeing `requestSite` patches they needed to do there.

jsnajdr committed:
requestSite should return a rejected promise on error

tyxla reviewed (commented):
This looks great, thank you! But we'll need to fix some failing tests that are affected by the changed behavior.

jsnajdr committed:
Fixup doRequest parameters

jsnajdr committed:
Fixup and modernize sites/actions tests

jsnajdr commented:
> But we'll need to fix some failing tests that are affected by the changed behavior.

Just pushed a few fixes. The `requestSite` action itself was wrong (forgotten `siteFragment` parameter) and the tests needed some update, too. I migrated them away from Chai and also removed `jest/no-conditional-expects` offenders (not reported and fixed by @scinos in #53206 because of Chai).

jsnajdr committed:
Fixup tests for Jetpack Authorize redirect

tyxla commented:
> > But we'll need to fix some failing tests that are affected by the changed behavior.
> 
> Just pushed a few fixes. The `requestSite` action itself was wrong (forgotten `siteFragment` parameter) and the tests needed some update, too. I migrated them away from Chai and also removed `jest/no-conditional-expects` offenders (not reported and fixed by @scinos in #53206 because of Chai).

Thanks, Jarda. I won't be able to get to it today but I'm planning to do some more thorough review and testing on Monday.

tyxla commented on client/my-sites/marketing/controller.js:
```diff
@@ -41,7 +41,11 @@ export const redirectDefaultConnectionsDomain = async ( context ) => {
+		try {
+			await dispatch( requestSite( recentSiteId ) );
+		} catch {
+			// proceed despite a failed site request
```
What's the idea here? To use the previously cached site data (if any)?

jsnajdr replied:
The idea is that the calling code doesn't check the return value of `requestSite`. Instead, it detects success or failure by looking at its side effects, i.e., what it wrote to state. Something like:
```js
if ( ! getSite( getState(), siteId ) ) {
  await dispatch( requestSite( siteId ) );
  if ( ! getSite( getState(), siteId ) ) {
    console.error( 'the site is not there even after requesting it, means the request failed' );
  }
}
```
The `.catch( () => null )` blocks in other parts of this patch express the same idea.

tyxla commented on client/state/sites/actions.js:
```diff
@@ -115,55 +115,52 @@ export function requestSites() {
 				// If we can't manage the site, don't add it to state.
-				if ( ! ( site && site.capabilities ) ) {
-					return dispatch( {
-						type: SITE_REQUEST_FAILURE,
```
I wonder if any of the consumer code was relying on the fact that we were dispatching a failure in the case when the user doesn't have enough permissions. I wasn't able to verify that at a single glance. Have you checked it yourself somehow?

jsnajdr replied:
Yes, I verified that -- the `SITE_REQUEST_FAILURE` and `SITE_REQUEST_SUCCESS` action only reset a `requesting` flag to `false`, they have no other function. They are identical to each other.

There is a special case in Jetpack Connect Authorize where the `FAILURE` action is used to set a `clientNotResponding` flag. But that component issues its own `wpcom.site().get()` request (in the `isUserConnected` action) and also dispatches its own `SITE_REQUEST_FAILURE` action. Changes in `state/sites` shouldn't affect that.

tyxla reviewed (commented):
This looks great, code-wise 👍  nice work!

Can I ask for some testing instructions, though? I did test the affected areas and didn't found any errors, but I wonder if I missed something in my testing.

I've also left a couple of minor questions.

jsnajdr committed:
Fix SiteAddressChanger crash when the .wordpress.com domain is not primary

jsnajdr commented:
> Can I ask for some testing instructions, though?

The main thing to test is the site selection for paths like `/home/example.wordpress.com`. In My Sites, in Jetpack Cloud, and also in signup when the path has a `?siteSlug=example.wordpress.com` query param. If the site slug is invalid, routes should redirect to all-sites paths.

I also tested changing site address in Domains Management for `*.wordpress.com` domains. And there I discovered a pre-existing bug: if the site has a mapped primary domain (`example.blog`) and we're changing the secondary `*.wordpress.com` domain, the address changer redirects to a wrong URL after the change is done. The target URL is in the form:
```
/domains/manage/:domain/edit/:site
```
and we should redirect to
```
/domains/manage/new.wordpress.com/edit/example.blog
```
I.e., use the new name of the domain and old (unchanged) site slug. But the address changer is redirecting to:
```
/domains/manage/new.wordpress.com/edit/new.wordpress.com
```
But `new.wordpress.com` is not the correct site slug. It's only a secondary URL of the site.

tyxla commented on client/state/site-address-change/actions.js:
```diff
@@ -126,7 +127,44 @@ export const requestSiteAddressChange = (
+			);
+
+			// site slug, potentially freshly updated by the `requestSite` above
+			const siteSlug = getSiteSlug( getState(), siteId );
```
Nice catch 👍  I wonder if we introduced this during the reduxification or it was a pre-existing bug.

jsnajdr replied:
I believe we broke this years ago when reduxifying `sites-list`. For a Simple site with a mapped domain, its "canonical" slug is `example.blog`, the mapped domain. The unmapped `example.wordpress.com` is an alternative slug. The Flux `sites-list`, when doing the "get site ID by slug" operation, supported both of them. But the Redux selector always supported only the `example.blog` one.

#53328 is actually a part of effort to finally fix this 🙂

tyxla reviewed (approved):
This looks great and tests well! Also, confirmed the domains fix - nice one! 

Good to go from my perspective.  🚢
